File: src/tsid_test_commander.cpp
#include <memory>
#include <vector>
#include <string>
#include <chrono>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "robosoccer_control/msg/tsid_command.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"

class TSIDTestCommander : public rclcpp::Node
{
public:
  TSIDTestCommander() : Node("tsid_test_commander")
  {
    // Declare parameters
    this->declare_parameter("test_mode", "standing");  // standing, com_tracking, foot_tracking
    this->declare_parameter("use_sim_time", false);
    this->declare_parameter("command_frequency", 10.0);  // Hz
    
    // Get parameters
    test_mode_ = this->get_parameter("test_mode").as_string();
    use_sim_time_ = this->get_parameter("use_sim_time").as_bool();
    command_frequency_ = this->get_parameter("command_frequency").as_double();

    // Publishers
    tsid_command_pub_ = this->create_publisher<robosoccer_control::msg::TSIDCommand>(
      "/tsid_commands", 10);
    
    com_target_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>(
      "/com_target", 10);
    
    left_foot_target_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>(
      "/left_foot_target", 10);
    
    right_foot_target_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>(
      "/right_foot_target", 10);

    // Control timer
    auto timer_period = std::chrono::duration<double>(1.0 / command_frequency_);
    command_timer_ = this->create_wall_timer(
      timer_period, std::bind(&TSIDTestCommander::commandLoop, this));

    // Initialize test state
    test_time_ = 0.0;
    phase_ = 0;

    RCLCPP_INFO(this->get_logger(), "TSID Test Commander initialized!");
    RCLCPP_INFO(this->get_logger(), "Test mode: %s", test_mode_.c_str());
    RCLCPP_INFO(this->get_logger(), "Command frequency: %.1f Hz", command_frequency_);
    RCLCPP_INFO(this->get_logger(), "Publishing to: /tsid_commands, /com_target, /left_foot_target, /right_foot_target");
  }

private:
  void commandLoop()
  {
    test_time_ += 1.0 / command_frequency_;
    
    if (test_mode_ == "standing") {
      sendStandingCommand();
    } else if (test_mode_ == "com_tracking") {
      sendCoMTrackingCommand();
    } else if (test_mode_ == "foot_tracking") {
      sendFootTrackingCommand();
    } else if (test_mode_ == "simple_motion") {
      sendSimpleMotionCommand();
    }
  }

  void sendStandingCommand()
  {
    // Send a standing command (CoM at center, feet in place)
    auto tsid_cmd = robosoccer_control::msg::TSIDCommand();
    tsid_cmd.header.stamp = this->now();
    
    // CoM at standing height
    tsid_cmd.desired_com_pose.position.x = 0.0;
    tsid_cmd.desired_com_pose.position.y = 0.0;
    tsid_cmd.desired_com_pose.position.z = 0.4;  // Standing height
    tsid_cmd.desired_com_pose.orientation.w = 1.0;
    
    // Zero velocity
    tsid_cmd.desired_com_velocity.linear.x = 0.0;
    tsid_cmd.desired_com_velocity.linear.y = 0.0;
    tsid_cmd.desired_com_velocity.linear.z = 0.0;
    
    // Feet in contact
    tsid_cmd.left_foot_contact = true;
    tsid_cmd.right_foot_contact = true;
    
    // Publish command
    tsid_command_pub_->publish(tsid_cmd);
    
    RCLCPP_INFO_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
                        "Sending standing command - CoM at (0, 0, 0.4)");
  }

  void sendCoMTrackingCommand()
  {
    // Send a CoM tracking command (circular motion)
    auto tsid_cmd = robosoccer_control::msg::TSIDCommand();
    tsid_cmd.header.stamp = this->now();
    
    // Circular CoM motion
    double radius = 0.05;  // 5cm radius
    double frequency = 0.5;  // 0.5 Hz
    
    tsid_cmd.desired_com_pose.position.x = radius * cos(2 * M_PI * frequency * test_time_);
    tsid_cmd.desired_com_pose.position.y = radius * sin(2 * M_PI * frequency * test_time_);
    tsid_cmd.desired_com_pose.position.z = 0.4;
    tsid_cmd.desired_com_pose.orientation.w = 1.0;
    
    // CoM velocity
    tsid_cmd.desired_com_velocity.linear.x = -radius * 2 * M_PI * frequency * sin(2 * M_PI * frequency * test_time_);
    tsid_cmd.desired_com_velocity.linear.y = radius * 2 * M_PI * frequency * cos(2 * M_PI * frequency * test_time_);
    tsid_cmd.desired_com_velocity.linear.z = 0.0;
    
    // Feet in contact
    tsid_cmd.left_foot_contact = true;
    tsid_cmd.right_foot_contact = true;
    
    // Publish command
    tsid_command_pub_->publish(tsid_cmd);
    
    RCLCPP_INFO_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
                        "Sending CoM tracking command - Circular motion");
  }

  void sendFootTrackingCommand()
  {
    // Send foot tracking commands
    auto left_foot_msg = geometry_msgs::msg::PoseStamped();
    auto right_foot_msg = geometry_msgs::msg::PoseStamped();
    
    left_foot_msg.header.stamp = this->now();
    right_foot_msg.header.stamp = this->now();
    
    // Simple foot motion (lifting one foot)
    if (phase_ < 100) {
      // Phase 1: Lift right foot
      right_foot_msg.pose.position.z = 0.05 * sin(M_PI * test_time_);
      left_foot_msg.pose.position.z = 0.0;
    } else if (phase_ < 200) {
      // Phase 2: Lift left foot
      right_foot_msg.pose.position.z = 0.0;
      left_foot_msg.pose.position.z = 0.05 * sin(M_PI * test_time_);
    } else {
      // Reset phase
      phase_ = 0;
    }
    
    // Publish foot targets
    left_foot_target_pub_->publish(left_foot_msg);
    right_foot_target_pub_->publish(right_foot_msg);
    
    phase_++;
    
    RCLCPP_INFO_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
                        "Sending foot tracking command - Phase %d", phase_);
  }

  void sendSimpleMotionCommand()
  {
    // Send a simple motion command (CoM forward/backward)
    auto tsid_cmd = robosoccer_control::msg::TSIDCommand();
    tsid_cmd.header.stamp = this->now();
    
    // Forward/backward motion
    double amplitude = 0.02;  // 2cm
    double frequency = 0.3;   // 0.3 Hz
    
    tsid_cmd.desired_com_pose.position.x = amplitude * sin(2 * M_PI * frequency * test_time_);
    tsid_cmd.desired_com_pose.position.y = 0.0;
    tsid_cmd.desired_com_pose.position.z = 0.4;
    tsid_cmd.desired_com_pose.orientation.w = 1.0;
    
    // CoM velocity
    tsid_cmd.desired_com_velocity.linear.x = amplitude * 2 * M_PI * frequency * cos(2 * M_PI * frequency * test_time_);
    tsid_cmd.desired_com_velocity.linear.y = 0.0;
    tsid_cmd.desired_com_velocity.linear.z = 0.0;
    
    // Feet in contact
    tsid_cmd.left_foot_contact = true;
    tsid_cmd.right_foot_contact = true;
    
    // Publish command
    tsid_command_pub_->publish(tsid_cmd);
    
    RCLCPP_INFO_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
                        "Sending simple motion command - Forward/backward");
  }

  rclcpp::Publisher<robosoccer_control::msg::TSIDCommand>::SharedPtr tsid_command_pub_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr com_target_pub_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr left_foot_target_pub_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr right_foot_target_pub_;
  
  rclcpp::TimerBase::SharedPtr command_timer_;
  
  std::string test_mode_;
  bool use_sim_time_;
  double command_frequency_;
  double test_time_;
  int phase_;
};

int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TSIDTestCommander>());
  rclcpp::shutdown();
  return 0;
} 

File: robosoccer_control/tsid_test_commander.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped
from robosoccer_control.msg import TSIDCommand
import math

class TSIDTestCommander(Node):
    def __init__(self):
        super().__init__('tsid_test_commander')
        self.declare_parameter('test_mode', 'standing')
        self.declare_parameter('command_frequency', 10.0)
        self.test_mode = self.get_parameter('test_mode').get_parameter_value().string_value
        self.command_frequency = self.get_parameter('command_frequency').get_parameter_value().double_value
        self.tsid_command_pub = self.create_publisher(TSIDCommand, '/tsid_commands', 10)
        self.com_target_pub = self.create_publisher(PoseStamped, '/com_target', 10)
        self.left_foot_target_pub = self.create_publisher(PoseStamped, '/left_foot_target', 10)
        self.right_foot_target_pub = self.create_publisher(PoseStamped, '/right_foot_target', 10)
        self.test_time = 0.0
        self.phase = 0
        self.create_timer(1.0 / self.command_frequency, self.command_loop)
        self.get_logger().info(f"TSID Test Commander initialized! Test mode: {self.test_mode}")

    def command_loop(self):
        self.test_time += 1.0 / self.command_frequency
        if self.test_mode == 'standing':
            self.send_standing_command()
        elif self.test_mode == 'com_tracking':
            self.send_com_tracking_command()
        elif self.test_mode == 'foot_tracking':
            self.send_foot_tracking_command()
        elif self.test_mode == 'simple_motion':
            self.send_simple_motion_command()

    def send_standing_command(self):
        tsid_cmd = TSIDCommand()
        tsid_cmd.desired_com_pose.position.x = 0.0
        tsid_cmd.desired_com_pose.position.y = 0.0
        tsid_cmd.desired_com_pose.position.z = 0.4
        tsid_cmd.desired_com_pose.orientation.w = 1.0
        tsid_cmd.desired_com_velocity.linear.x = 0.0
        tsid_cmd.desired_com_velocity.linear.y = 0.0
        tsid_cmd.desired_com_velocity.linear.z = 0.0
        tsid_cmd.left_foot_contact = True
        tsid_cmd.right_foot_contact = True
        self.tsid_command_pub.publish(tsid_cmd)
        self.get_logger().info_throttle(2.0, "Sending standing command - CoM at (0, 0, 0.4)")

    def send_com_tracking_command(self):
        tsid_cmd = TSIDCommand()
        radius = 0.05
        frequency = 0.5
        tsid_cmd.desired_com_pose.position.x = radius * math.cos(2 * math.pi * frequency * self.test_time)
        tsid_cmd.desired_com_pose.position.y = radius * math.sin(2 * math.pi * frequency * self.test_time)
        tsid_cmd.desired_com_pose.position.z = 0.4
        tsid_cmd.desired_com_pose.orientation.w = 1.0
        tsid_cmd.desired_com_velocity.linear.x = -radius * 2 * math.pi * frequency * math.sin(2 * math.pi * frequency * self.test_time)
        tsid_cmd.desired_com_velocity.linear.y = radius * 2 * math.pi * frequency * math.cos(2 * math.pi * frequency * self.test_time)
        tsid_cmd.desired_com_velocity.linear.z = 0.0
        tsid_cmd.left_foot_contact = True
        tsid_cmd.right_foot_contact = True
        self.tsid_command_pub.publish(tsid_cmd)
        self.get_logger().info_throttle(2.0, "Sending CoM tracking command - Circular motion")

    def send_foot_tracking_command(self):
        left_foot_msg = PoseStamped()
        right_foot_msg = PoseStamped()
        if self.phase < 100:
            right_foot_msg.pose.position.z = 0.05 * math.sin(math.pi * self.test_time)
            left_foot_msg.pose.position.z = 0.0
        elif self.phase < 200:
            right_foot_msg.pose.position.z = 0.0
            left_foot_msg.pose.position.z = 0.05 * math.sin(math.pi * self.test_time)
        else:
            self.phase = 0
        self.left_foot_target_pub.publish(left_foot_msg)
        self.right_foot_target_pub.publish(right_foot_msg)
        self.phase += 1
        self.get_logger().info_throttle(2.0, f"Sending foot tracking command - Phase {self.phase}")

    def send_simple_motion_command(self):
        tsid_cmd = TSIDCommand()
        amplitude = 0.02
        frequency = 0.3
        tsid_cmd.desired_com_pose.position.x = amplitude * math.sin(2 * math.pi * frequency * self.test_time)
        tsid_cmd.desired_com_pose.position.y = 0.0
        tsid_cmd.desired_com_pose.position.z = 0.4
        tsid_cmd.desired_com_pose.orientation.w = 1.0
        tsid_cmd.desired_com_velocity.linear.x = amplitude * 2 * math.pi * frequency * math.cos(2 * math.pi * frequency * self.test_time)
        tsid_cmd.desired_com_velocity.linear.y = 0.0
        tsid_cmd.desired_com_velocity.linear.z = 0.0
        tsid_cmd.left_foot_contact = True
        tsid_cmd.right_foot_contact = True
        self.tsid_command_pub.publish(tsid_cmd)
        self.get_logger().info_throttle(2.0, "Sending simple motion command - Forward/backward")

def main(args=None):
    rclpy.init(args=args)
    node = TSIDTestCommander()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main() 